feat: add guide and api docs for react lynx testing library (#117)
lynx-family/lynx-website: docs/en/react/react-lynx-testing-library/index.mdx

## Applicable scenarios

ReactLynx Testing Library is suitable for unit testing of ReactLynx components. It is based on Lynx Environment and `@testing-library/dom` (querying and event triggering of page elements), shielding the implementation details of Lynx dual threads, and abstracting the rendering process of ReactLynx into `render`, `fireEvent`, `screen` and other APIs similar to React Testing Library. Users can also assert on the elements in the page, for example, use `data-testid` to quickly find the element, and use `toBeInTheDocument` in `@testing-library/jest-dom` to determine whether the element is in the page.

## Installation

<PackageManagerTabs command="add -D @lynx-js/react-lynx-testing-library vitest" />

Configure Vitest:

```js title=vitest.config.js
import { defineConfig, mergeConfig } from 'vitest/config';
import { createVitestConfig } from '@lynx-js/react-lynx-testing-library/vitest-config';

const defaultConfig = createVitestConfig();
const config = defineConfig({
  test: {
    // ...
  },
});

export default defineConfig(mergeConfig(defaultConfig, config));
```

Then you can start writing tests and running them!

> :::info No need to install @lynx-js/environment

Reviewer: I personally feel a little bit strange reading here...

Do you want to emphasize "no need to install" or "although we are based on @lynx-js/environment, we have already packaged it for you"? If it is the latter, you can wrap this part with `<Details/>`?

```tsx
import { Details } from '@lynx';

<Details>
...
</Details>
```

  expect(elementTree.root).toMatchInlineSnapshot(`
    <page>
      <text>
        Loading...
      </text>
    </page>
  `);
  const loading = () => {
    return screen.getByText('Loading...');
  };
  await waitForElementToBeRemoved(loading);
  // Since Lynx Environment uses jsdom to implement Element PAPI at the bottom layer
  // you can directly access document.body to get page elements
  expect(document.body).toMatchInlineSnapshot(`
    <body>
      <page>
        <text
          data-testid="message"
        >
          Loaded this message:
          <wrapper>
            Hello World
          </wrapper>
          !
        </text>
      </page>
    </body>
  `);
  expect(screen.getByTestId('message')).toHaveTextContent(/Hello World/);
  expect(elementTree.root).toMatchInlineSnapshot(`
    <page>
      <text
        data-testid="message"
      >
        Loaded this message:
        <wrapper>
          Hello World
        </wrapper>
        !
      </text>
    </page>
  `);
});
```

### Rerender

```jsx
import '@testing-library/jest-dom';
import { render } from '@lynx-js/react-lynx-testing-library';
import { expect } from 'vitest';

> it('rerender will re-render the element', async () => {

Reviewer: nit
```suggestion
it('rerender will re-render your component', async () => {
```

  const Greeting = (props) => <text>{props.message}</text>;
  const { container, rerender } = render(<Greeting message="hi" />);
  expect(container).toMatchInlineSnapshot(`
    <page>
      <text>
        hi
      </text>
    </page>
  `);
  expect(container.firstChild).toHaveTextContent('hi');

  {
    // Unlike React Testing Library, container needs to be retrieved after rerender

>     // Because ReactLynx creates a new page element each time it loads

Reviewer: I can't quite remember but we created a module-scoped persistent page and stored it in a `__page` variable?

    const { container } = rerender(<Greeting message="hey" />);
    expect(container.firstChild).toHaveTextContent('hey');

    expect(container).toMatchInlineSnapshot(`
      <page>
        <text>
          hey
        </text>
      </page>
    `);
  }
});
```

> ### More usage

Reviewer: `<list/>` will be in "More usage", or should we guide here?

Author: I have added the list and mts example, PTAL again~

            <text>{item}</text>
          </list-item>
        ))}
      </list>
    );
  };
  const { container } = render(<Comp />);
  expect(container).toMatchInlineSnapshot(`
    <page>
      <list
        update-list-info="[{"insertAction":[{"position":0,"type":"__Card__:__snapshot_f75b7_test_2","item-key":0},{"position":1,"type":"__Card__:__snapshot_f75b7_test_2","item-key":1},{"position":2,"type":"__Card__:__snapshot_f75b7_test_2","item-key":2}],"removeAction":[],"updateAction":[]}]"
      />
    </page>
  `);
  const list = container.firstChild;
  // Trigger the componentAtIndex method of list and load the 0th item
  const uid0 = elementTree.enterListItemAtIndex(list, 0);
  expect(list).toMatchInlineSnapshot(`
    <list
      update-list-info="[{"insertAction":[{"position":0,"type":"__Card__:__snapshot_f75b7_test_2","item-key":0},{"position":1,"type":"__Card__:__snapshot_f75b7_test_2","item-key":1},{"position":2,"type":"__Card__:__snapshot_f75b7_test_2","item-key":2}],"removeAction":[],"updateAction":[]}]"
    >
      <list-item
        item-key="0"
      >
        <text>
          0
        </text>
      </list-item>
    </list>
  `);

>   // Trigger the enqueueComponent method of list to notify that 0th item is recycled

Reviewer: nit: The comment is inconsistent with code below

      <page>
        <text>
          hey
        </text>
      </page>
    `);
  }
});
```

### list

> For list usage documentation, please refer to: https://lynxjs.org/zh/api/elements/built-in/list.html

```jsx
import { useState } from '@lynx-js/react';
import { render } from '@lynx-js/react/testing-library';
import { expect } from 'vitest';

it('list', () => {
  const Comp = () => {
    const [list, setList] = useState([0, 1, 2]);
    return (
      <list>
        {list.map((item) => (
          <list-item key={item} item-key={item}>
            <text>{item}</text>
          </list-item>
        ))}
      </list>
    );
  };
  const { container } = render(<Comp />);
  expect(container).toMatchInlineSnapshot(`
    <page>
      <list
        update-list-info="[{"insertAction":[{"position":0,"type":"__Card__:__snapshot_f75b7_test_2","item-key":0},{"position":1,"type":"__Card__:__snapshot_f75b7_test_2","item-key":1},{"position":2,"type":"__Card__:__snapshot_f75b7_test_2","item-key":2}],"removeAction":[],"updateAction":[]}]"
      />
    </page>
  `);
  const list = container.firstChild;
  // Trigger the componentAtIndex method of list and load the 0th item

>   const uid0 = elementTree.triggerComponentAtIndex(list, 0);

Reviewer: Before we bring these internal api into public, can we re-consider the naming of it?

At least replace `Component` with `listItem` or something similar :D

Author: - `triggerComponentAtIndex` calls `componentAtIndex` of list element
- `triggerEnqueueComponent` calls `enqueueComponent` of list element

However, `componentAtIndex` and `enqueueComponent` are not directly exposed to users.

Let me use `enterListItemAtIndex` and `leaveListItem` for now.

https://github.com/upupming/lynx-stack/blob/97557eb25c3466a6fba8a28713155fa621e9de4f/packages/testing-library/lynx-environment/src/lynx/ElementPAPI.ts#L437

  expect(cb.mock.calls).toMatchInlineSnapshot(`
    [
      [
        {
          "eventName": "tap",
          "eventType": "bindEvent",
          "isTrusted": false,
          "key": "value",
        },
      ],
    ]
  `);
});
```

#### Choose the thread used for rendering

The second parameter of the `render` method supports `enableMainThread` and `enableBackgroundThread` are two options for enabling the main thread and background threads. By default, `enableMainThread` is `false` and `enableBackgroundThread` is `true`, so the main thread does not render the first screen, only the background thread runs the full Preact runtime and sends UI updates to the main thread for rendering. The reason why we don't enable both the main thread and the background thread is because unit tests do not compile files into two products like using Rspeedy does, which will result in the top-level code being executed only once, and by default in the background thread. Suppose you wrote a component like this:

```jsx
const isBackground = __BACKGROUND__;
const Comp = () => {
  return (
    <view>
      <text>{isBackground ? 'background' : 'main thread'}</text>
    </view>
  );
};
```

> The top-level `isBackground` of this component will only be executed once, and by default it is `true` (Lynx Environment defaults to the background thread after initialization). In order to avoid rendering the wrong first screen result when `enableMainThread: true`, we set `enableMainThread` to `false` by default, and get an empty first screen, which can avoid this problem.

Reviewer: nit:
```suggestion
The top-level `isBackground` of this module will only be executed once, and by default it is `true` (Lynx Environment defaults to the background thread after initialization). In order to avoid rendering the wrong first screen result when `enableMainThread: true`, we set `enableMainThread` to `false` by default, and get an empty first screen, which can avoid this problem.
```
